src/transformers/models/ict/modeling_ict.py

```diff
@@ -0,0 +1,904 @@
+
+        # TODO How to load weights for G and D?
+        # if len(config.GPU) > 1:
+        #     generator = nn.DataParallel(generator, config.GPU)
+        #     discriminator = nn.DataParallel(discriminator , config.GPU)
+
+        l1_loss = nn.L1Loss()
+        perceptual_loss = PerceptualLoss()
+        style_loss = StyleLoss()
+        adversarial_loss = AdversarialLoss(type=config.gan_loss)
+
+        self.add_module("generator", generator)
+        self.add_module("discriminator", discriminator)
+
+        self.add_module("l1_loss", l1_loss)
+        self.add_module("perceptual_loss", perceptual_loss)
+        self.add_module("style_loss", style_loss)
+        self.add_module("adversarial_loss", adversarial_loss)
+
+        self.gen_optimizer = optim.Adam(
+            params=generator.parameters(), lr=float(config.learning_rate), betas=(config.adam_beta1, config.adam_beta2)
+        )
+
+        self.dis_optimizer = optim.Adam(
+            params=discriminator.parameters(),
+            lr=float(config.learning_rate) * float(config.dis_gen_learning_rate),
+            betas=(config.adam_beta1, config.adam_beta2),
+        )
+
+    def process(self, images, edges, masks):
+        self.iteration += 1
+
+        # zero optimizers
+        self.gen_optimizer.zero_grad()
+        self.dis_optimizer.zero_grad()
+
+        # process outputs
+        outputs = self(images, edges, masks)
+        gen_loss = 0
+        dis_loss = 0
+
+        # discriminator loss
+        dis_input_real = images
+        dis_input_fake = outputs.detach()
+        dis_real, _ = self.discriminator(dis_input_real)
+        dis_fake, _ = self.discriminator(dis_input_fake)
+        dis_real_loss = self.adversarial_loss(dis_real, True)
+        dis_fake_loss = self.adversarial_loss(dis_fake, False)
+        dis_loss += (dis_real_loss + dis_fake_loss) / 2
+
+        dis_loss.backward()
+        self.dis_optimizer.step()
+
+        # generator adversarial loss
+        gen_input_fake = outputs
+        gen_fake, _ = self.discriminator(gen_input_fake)
+        gen_gan_loss = self.adversarial_loss(gen_fake, True) * self.config.INPAINT_ADV_LOSS_WEIGHT
+        gen_loss += gen_gan_loss
+
+        # generator l1 loss
+        gen_l1_loss = self.l1_loss(outputs, images) * self.config.L1_LOSS_WEIGHT / torch.mean(masks)
+        gen_loss += gen_l1_loss
+
+        # generator perceptual loss
+        gen_content_loss = self.perceptual_loss(outputs, images)
+        gen_content_loss = gen_content_loss * self.config.CONTENT_LOSS_WEIGHT
+        gen_loss += gen_content_loss
+
+        # generator style loss
+        gen_style_loss = self.style_loss(outputs * masks, images * masks)
+        gen_style_loss = gen_style_loss * self.config.STYLE_LOSS_WEIGHT
+        gen_loss += gen_style_loss
+
+        gen_loss.backward()
+        self.gen_optimizer.step()
+
+        return outputs, gen_loss, dis_loss
```

Comment:
Since these are used for training, I was wondering how these can be used when training?